opm/simulators/wells/BlackoilWellModelConstraints.cpp

```diff
@@ -0,0 +1,479 @@
+
+            // sum over all nodes
+            current_rate = well_.comm().sum(current_rate);
+
+            if (controls.water_target < current_rate  ) {
+                double scale = 1.0;
+                if (current_rate > 1e-12)
+                    scale = controls.water_target / current_rate;
+                return std::make_pair(Group::ProductionCMode::WRAT, scale);
+            }
+        }
+    }
+    if (group.has_control(Group::ProductionCMode::GRAT))
+    {
+        if (currentControl != Group::ProductionCMode::GRAT)
+        {
+            double current_rate = 0.0;
+            current_rate += WellGroupHelpers::sumWellSurfaceRates(group, well_.schedule(), well_state, reportStepIdx, pu.phase_pos[BlackoilPhases::Vapour], false);
+
+            // sum over all nodes
+            current_rate = well_.comm().sum(current_rate);
+            if (controls.gas_target < current_rate  ) {
+                double scale = 1.0;
+                if (current_rate > 1e-12)
+                    scale = controls.gas_target / current_rate;
+                return std::make_pair(Group::ProductionCMode::GRAT, scale);
+            }
+        }
+    }
+    if (group.has_control(Group::ProductionCMode::LRAT))
+    {
+        if (currentControl != Group::ProductionCMode::LRAT)
+        {
+            double current_rate = 0.0;
+            current_rate += WellGroupHelpers::sumWellSurfaceRates(group, well_.schedule(), well_state, reportStepIdx, pu.phase_pos[BlackoilPhases::Liquid], false);
+            current_rate += WellGroupHelpers::sumWellSurfaceRates(group, well_.schedule(), well_state, reportStepIdx, pu.phase_pos[BlackoilPhases::Aqua], false);
+
+            // sum over all nodes
+            current_rate = well_.comm().sum(current_rate);
+
+            bool skip = false;
+            if (controls.liquid_target == controls.oil_target) {
+                double current_water_rate = WellGroupHelpers::sumWellSurfaceRates(group, well_.schedule(), well_state, reportStepIdx, pu.phase_pos[BlackoilPhases::Aqua], false);
+                current_water_rate = well_.comm().sum(current_water_rate);
+                if (std::abs(current_water_rate) < 1e-12) {
+                    skip = true;
+                    deferred_logger.debug("LRAT_ORAT_GROUP", "GROUP " + group.name() + " The LRAT target is equal the ORAT target and the water rate is zero, skip checking LRAT");
+                }
+            }
+
+            if (!skip && controls.liquid_target < current_rate ) {
+                double scale = 1.0;
+                if (current_rate > 1e-12)
+                    scale = controls.liquid_target / current_rate;
+                return std::make_pair(Group::ProductionCMode::LRAT, scale);
+            }
+        }
+    }
+
+    if (group.has_control(Group::ProductionCMode::CRAT))
+    {
+        OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "CRAT control for production groups not implemented" , deferred_logger);
```

Comment:
We should include `<stdexcept>` to bring a declaration of `std::runtime_error` into scope.

Reply:
done.